### PR TITLE
unblock Unix Domain Socket on Dispose on macOS

### LINKED ISSUE #101539: Unable to cancel NamedPipeClientStream reads on MacOS
### Description

I am unable to cancel reads on `NamedPipeClientStream` on MacOS, as it seems to deadlock when closing the reader.

### Reproduction Steps

The following code reproduces the problem on MacOS, but works fine on both Windows and Linux:

```cs
using System.IO.Pipes;

new Thread(() => {
	using var host = new NamedPipeServerStream(
		"cstest_pipes",
		PipeDirection.InOut,
		NamedPipeServerStream.MaxAllowedServerInstances,
		PipeTransmissionMode.Byte,
		PipeOptions.Asynchronous
	);

	Console.WriteLine("Host: Waiting for connection..");
	host.WaitForConnection();

	Console.WriteLine("Host: Connected :)");

	while (true) {
		Thread.Sleep(1000);
	}
}).Start();

using var client = new NamedPipeClientStream(".", "cstest_pipes", PipeDirection.InOut, PipeOptions.Asynchronous);

Console.WriteLine("Client: Connecting..");
client.Connect(TimeSpan.FromSeconds(1));

Console.WriteLine($"Client: Connected: {client.IsConnected}");

var cts = new CancellationTokenSource();
var ct = cts.Token;

using var reader = new BinaryReader(client);
using var readerReg = ct.Register(reader.Close);

new Task(async () => {
	Console.WriteLine("Client: Cancelling in 500ms..");
	await Task.Delay(TimeSpan.FromMilliseconds(500));

	Console.WriteLine("Client: Cancelling..");
	cts.Cancel();

	Console.WriteLine("Client: Cancelled :)");
}).Start();

while (!ct.IsCancellationRequested) {
	Console.WriteLine("Client: Reading byte..");
	try {
		var b = reader.ReadByte();
		Console.WriteLine($"Client: Byte received: 0x{b:02x}");
	} catch (Exception) {
		Console.WriteLine("Client: Exception caught while reading byte");
		break;
	}
}

Console.WriteLine("Client: Disconnected :)");
```

### Expected behavior

On stream close, I expect any existing reads to be cancelled and throw an exception. In other words, I expect the output of the above program to be something like this:

```
Client: Connecting..
Host: Waiting for connection..
Client: Connected: True
Client: Reading byte..
Host: Connected :)
Client: Cancelling in 500ms..
Client: Cancelling..
Client: Exception caught while reading byte
Client: Cancelled :)
Client: Disconnected :)
```

### Actual behavior

In this case, a deadlock occurs between the stream closing and the stream being read from. In the example code above, the output of the program locks up here:

```
Host: Waiting for connection..
Client: Connected: True
Client: Reading byte..
Host: Connected :)
Client: Cancelling in 500ms..
Client: Cancelling..
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

Tested with SDK `9.0.100-preview.3.24204.13` (it happens on `8.0.203` as well)
MacOS Sonoma 14.4.1, arm64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by wfurt**
fixes #101539
When. `Interop.Sys.Disconnect` is called on macOS with UDS, it fails and we end up in endless `TryUnblockSocket` loop. 

We _could_ fallback to `Shutdown` on  any error - this is just smallest change to fix the reported issue., 
Lot of the abortive jargon around here really make sense only for TCP imho and I'm n to sure how applicable the logic is to other socket types.

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs [hunk 1]
```diff
@@ -207,7 +207,10 @@ private unsafe bool TryUnblockSocket(bool abortive)
                 // Unless the user requested a normal close using Socket.Shutdown.
                 if (type == (int)SocketType.Stream && !_hasShutdownSend)
                 {
-                    Interop.Sys.Disconnect(handle);
+                    if (Interop.Sys.Disconnect(handle) == Interop.Error.EOPNOTSUPP)
+                    {
+                        Interop.Sys.Shutdown(handle, SocketShutdown.Both);
+                    }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:213]: Could this instead be:
```C#
if (type != (int)SocketType.Stream ||
    _hasShutdownSend ||
    Interop.Sys.Disconnect(handle) != Interop.Error.EOPNOTSUPP)
{
    Interop.Sys.Shutdown(handle, SocketShutdown.Both);
}
```
?

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:213]: right. Right now we have  somewhat distinctive path stream vs datagram. And abortive vs non abortive in some other cases. We can compress the code but it _may_ Loose some clarity.  Maybe some comments? 
Maybe @tmds can chime in?

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:213]: Whichever you prefer

**Comment by wfurt** [comment]
Note that the  linked issue can also be fixed by adding `Shutdown` to NamedPipeClientStream's  `Dispose` to make sure  everything is written and disconnected.  (or we dan perhaps do both)

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs (hunk 1 above)
Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:213]: `SystemNative_Disconnect` already uses/falls back to `shutdown`. We can move this change there by also using `shutdown` as a fallback on systems that `HAVE_DISCONNECTX`.

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs]: updated.

Independently, neither of the calls update ` _hasShutdownSend` so it seems like we can call it over and over again. Is there some benefit to it @tmds?

Comment (tmds) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs]: The reason for calling it again is for dealing with a race that might happen if an operation has already reffed the handle before we called `Dispose` and then starts the operation using that handle after we've called `TryUnblockSocket`. The next call to `TryUnblockSocket` is responsible for unblocking that operation (if it was blocking one).

With  `_hasShutdownSend` we're tracking if the user has explicitly called shutdown. If he did, we want the peer to see a regular TCP close.

File: src/native/libs/System.Native/pal_networking.c [hunk 2]
```diff
@@ -3143,6 +3143,11 @@ int32_t SystemNative_Disconnect(intptr_t socket)
 #elif HAVE_DISCONNECTX
     // disconnectx causes a FIN close on OSX. It's the best we can do.
     err = disconnectx(fd, SAE_ASSOCID_ANY, SAE_CONNID_ANY);
+    if (err != 0 && errno == EOPNOTSUPP)
```
Comment (tmds) [inline src/native/libs/System.Native/pal_networking.c:3146]: We could use the same condition here like we do with Linux: `if (err != 0)`.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs [hunk 3]
```diff
@@ -586,6 +586,54 @@ public async Task ReceiveFrom_EndPoints_Correct(bool useAsync)
...
+                    {
+                        Task disposeTask = Task.Run(() =>
+                        {
+                            Thread.Sleep(msDelay);
+                            client.Dispose();
+                        });
+
+                        try
+                        {
+                            client.Receive(buffer);
```
Comment (tmds) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs:618]: If there's a regression that causes `TryUnblockSocket` to break, the test will hang at `client.Receive` and never time out.

Comment (tmds) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs:618]: We can do some things like:

https://github.com/dotnet/runtime/blob/5d0536142062ffe5d1222b02f4ef1943f878cb4a/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs#L1079-L1082

Comment (wfurt) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs:618]: I was thinking about it. However, I think we would notice it and fix it. XUnit now spits out warnings about long running tasks.

File: src/native/libs/System.Native/pal_networking.c (hunk 2 above)
Comment (wfurt) [reply src/native/libs/System.Native/pal_networking.c]: updated

**Review by tmds — APPROVED** [review]

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs [hunk 4]
```diff
@@ -586,6 +586,54 @@ public async Task ReceiveFrom_EndPoints_Correct(bool useAsync)
...
+
+                client.Connect(endPoint);
+                int msDelay = 100;
+                bool readFailed = false;
+                byte[] buffer = new byte[100];
+                using (Socket accepted = server.Accept())
+                {
+                    while (msDelay < 10_000)
+                    {
+                        Task disposeTask = Task.Run(() =>
```
Comment (danmoseley) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs:610]: Are you waiting on this somewhere?

Comment (wfurt) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/UnixDomainSocketTest.cs:610]: it should not matter. But I added await just the clean up the task properly.

**Review by MihaZupan — APPROVED** [review]

**Review by antonfirsov — APPROVED** [review]

**Pull request merged**